### PR TITLE
PublicKeyValue stringer fix

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/dapperlabs/flow-internal/issues/1515

## Description

After the `PublicKeyValue.publicKey` field was changed to a computed field (https://github.com/onflow/cadence/pull/879), it no longer appeared if logged (`.String()`). It is desired that the public key is still a part of  `.String()` output.

There is also a test suite in the FVM that relies on this: https://github.com/onflow/flow-go/blob/4c6b6f6916d8f9d3ce347f946e6db1b84ed77c89/fvm/accounts_test.go#L1007

I added a custom `PublicKeyValue` stringer that includes the `publicKey` field.

Because stringers were not copied when a composite was copied I also included that change.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/value.go [hunk 1]
```diff
@@ -6173,6 +6173,7 @@ func (v *CompositeValue) Copy() Value {
 		// NOTE: new value has no owner
 		Owner:    nil,
 		modified: true,
+		stringer: v.stringer,
```
Comment (SupunS) [inline runtime/interpreter/value.go:6176]: Thanks for catching this!

File: runtime/interpreter/value.go [hunk 2]
```diff
@@ -8179,6 +8180,17 @@ func NewPublicKeyValue(
 	// Validate the public key, and initialize 'isValid' field.
 	publicKeyValue.Fields.Set(sema.PublicKeyIsValidField, validationFunction(publicKeyValue))
 
+	// Public key value to string should include the key even though it is a computed field
+	stringerFields := NewStringValueOrderedMap()
+	stringerFields.Set(sema.PublicKeyPublicKeyField, publicKey.Copy())
+	publicKeyValue.Fields.Foreach(func(key string, value Value) {
+		stringerFields.Set(key, value)
+	})
```
Comment (SupunS) [inline runtime/interpreter/value.go:8188]: :+1:

**Review by SupunS — APPROVED** [review]
Looks good! Thanks, @janezpodhostnik for the fix!

File: runtime/interpreter/value.go [hunk 3]
```diff
@@ -8179,6 +8180,17 @@ func NewPublicKeyValue(
...
+	stringer := func() string {
+		return formatComposite(string(publicKeyValue.TypeID()), stringerFields)
+	}
+	publicKeyValue.stringer = stringer
```
Comment (turbolent) [inline runtime/interpreter/value.go:8192]: Could we maybe delay the construction of the map required for the formatting until it is actually is needed? Also, I think the `ForEach` can be inlined. Maybe something like this:

```suggestion
	// Public key value to string should include the key even though it is a computed field
	var *stringerFields *StringValueOrderedMap
	publicKeyValue.stringer = func() string {
		if stringerFields == nil {
			stringerFields = NewStringValueOrderedMap()
			stringerFields.Set(sema.PublicKeyPublicKeyField, publicKey.Copy())
			publicKeyValue.Fields.Foreach(stringerFields.Set)
		}
		return formatComposite(string(publicKeyValue.TypeID()), stringerFields)
	}
```

File: runtime/interpreter/value_test.go [hunk 4]
```diff
@@ -366,13 +367,17 @@ func TestSetOwnerCompositeCopy(t *testing.T) {
...
+		return "random string"
+	}
 
 	compositeCopy := composite.Copy().(*CompositeValue)
 	valueCopy, _ := compositeCopy.Fields.Get(fieldName)
 
 	assert.Nil(t, compositeCopy.GetOwner())
 	assert.Nil(t, valueCopy.GetOwner())
 	assert.Equal(t, &oldOwner, value.GetOwner())
+	assert.Equal(t, composite.stringer(), compositeCopy.stringer())
```
Comment (turbolent) [inline runtime/interpreter/value_test.go:380]: Good catch, thanks for fixing and testing this 👍

**Review by turbolent — APPROVED** [review]

**Comment by turbolent** [comment]
Thank you for adding this @janezpodhostnik 👍

File: runtime/interpreter/value.go (hunk 3 above)
Comment (janezpodhostnik) [reply runtime/interpreter/value.go:8192]: stringerFields.Set is a `func (key string, value Value) (oldValue Value, present bool)` Foreach expects `func (key string, value Value)` so inlining like that wont work

Comment (turbolent) [reply runtime/interpreter/value.go]: Ah, right! 👍

File: runtime/interpreter/value.go [hunk 5]
```diff
@@ -6241,6 +6241,7 @@ func (v *CompositeValue) Copy() Value {
 		// NOTE: new value has no owner
 		Owner:           nil,
 		modified:        true,
+    stringer: v.stringer,
```
Comment (turbolent) [inline runtime/interpreter/value.go:6244]:
```suggestion
		stringer: v.stringer,
```

File: runtime/interpreter/value.go [hunk 6]
```diff
@@ -6241,6 +6241,7 @@ func (v *CompositeValue) Copy() Value {
 		// NOTE: new value has no owner
 		Owner:           nil,
 		modified:        true,
+		stringer: v.stringer,
```
Comment (turbolent) [inline runtime/interpreter/value.go:6244]:
```suggestion
		stringer:        v.stringer,
```

**Review by turbolent — APPROVED** [review]

**Comment by turbolent** [comment]
@janezpodhostnik resolved the conflicts, ready to be merged

File: runtime/interpreter/value_test.go [hunk 7]
```diff
@@ -2089,3 +2097,46 @@ func TestNumberValue_Equal(t *testing.T) {
...
+
+		sigAlgo := func() *CompositeValue {
+			fields := NewStringValueOrderedMap()
+			fields.Set(sema.EnumRawValueFieldName, UInt8Value(sema.SignatureAlgorithmECDSA_secp256k1.RawValue()))
+
+			return &CompositeValue{
+				qualifiedIdentifier: sema.SignatureAlgorithmType.QualifiedIdentifier(),
+				kind:                sema.SignatureAlgorithmType.Kind,
+				fields:              fields,
+			}
```
Comment (turbolent) [inline runtime/interpreter/value_test.go:2125]: @SupunS is this construction of a non-deferred composite value here correct?

**Review by turbolent — APPROVED** [review]

File: runtime/interpreter/value_test.go (hunk 7 above)
Comment (SupunS) [reply runtime/interpreter/value_test.go:2125]: This looks OK. We could also use the constructor, but that takes additional two params (owner and location).

**Review by SupunS — APPROVED** [review]

**Pull request merged**